include/votca/tools/graph.h

```diff
@@ -0,0 +1,126 @@
+/*
+ *            Copyright 2009-2018 The VOTCA Development Team
+ *                       (http://www.votca.org)
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *              http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <iostream>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include <votca/tools/edgecontainer.h>
+#include <votca/tools/graphnode.h>
+#include <votca/tools/name.h>
+
+#ifndef _VOTCA_TOOLS_GRAPH_H
+#define _VOTCA_TOOLS_GRAPH_H
+
+namespace votca {
+namespace tools {
+
+/**
+ * \brief A graph object that contains the graph nodes and the edges describing
+ *        the bonds between nodes.
+ *
+ */
+
+class GraphNode;
+
+class Graph : public EdgeContainer {
+ private:
+  /// Parameter description 
+  /// @param int - is the index of the graph nodes / vertex ids
+  /// @param GraphNode - this is the node object at each vertex and contains 
+  /// all the informatino that is relevant to that object
+  std::unordered_map<int, GraphNode> nodes_;
+
+  /// This is the id of the graph to graphs that contain the same content 
+  /// are considered equal
+  std::string id_;
+
+  /// If the id has been calculated yet
+  bool id_set_;
+  /// Update the graphs ids
+  void updateIds_(Graph& g);
```

Comment:
what does this method do? is it updating `this` graph id or it is a static method that takes any Graph and updates `id` ?

Reply:
Ya, I guess it is not very clear, it is used before the graphs are compared with "!=" and "==" the id is what is used to determine if the graphs are equal. The id is a string that is composed based on the sorted contents of each of the graph nodes. Thus two graphs will be equal if their graph nodes are the same. Because it is costly to create the id it is made once and only updated if a graph node has been changed or if it has not yet been set.

Reply:
I think perhaps it would be clearer if it only updated it's own id.